File: packages/go/vc-init.go
// vc-init.go - Bootstrap wrapper for standalone Go servers on Vercel
// This handles the IPC protocol required for executable runtime mode.
//
// The bootstrap:
// 1. Connects to VERCEL_IPC_PATH Unix socket
// 2. Starts the user's server on an internal port
// 3. Sends "server-started" IPC message
// 4. Reverse proxies requests to user's server
// 5. Handles /_vercel/ping health check
// 6. Sends "end" IPC message after each request

package main

import (
	"context"
	"encoding/json"
	"fmt"
	"net"
	"net/http"
	"net/http/httputil"
	"net/url"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"sync"
	"time"
)

// IPC message types
type StartMessage struct {
	Type    string       `json:"type"`
	Payload StartPayload `json:"payload"`
}

type StartPayload struct {
	InitDuration int `json:"initDuration"`
	HTTPPort     int `json:"httpPort"`
}

type EndMessage struct {
	Type    string     `json:"type"`
	Payload EndPayload `json:"payload"`
}

type EndPayload struct {
	Context RequestContext `json:"context"`
	Error   interface{}    `json:"error,omitempty"`
}

type RequestContext struct {
	InvocationID string `json:"invocationId"`
	RequestID    uint64 `json:"requestId"`
}

var (
	ipcConn   net.Conn
	ipcMutex  sync.Mutex
	ipcReady  bool
	startTime time.Time
)

func sendIPCMessage(msg interface{}) error {
	if ipcConn == nil {
		return nil
	}

	ipcMutex.Lock()
	defer ipcMutex.Unlock()

	data, err := json.Marshal(msg)
	if err != nil {
		return err
	}

	// IPC messages are JSON followed by null byte
	_, err = ipcConn.Write(append(data, 0))
	return err
}

func connectIPC() error {
	ipcPath := os.Getenv("VERCEL_IPC_PATH")
	if ipcPath == "" {
		// No IPC path - running in dev mode or locally
		return nil
	}

	conn, err := net.Dial("unix", ipcPath)
	if err != nil {
		return fmt.Errorf("failed to connect to IPC socket: %w", err)
	}

	ipcConn = conn
	return nil
}

func findFreePort() (int, error) {
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		return 0, err
	}
	port := listener.Addr().(*net.TCPAddr).Port
	listener.Close()
	return port, nil
}

func waitForServer(port int, timeout time.Duration) error {
	deadline := time.Now().Add(timeout)
	url := fmt.Sprintf("http://127.0.0.1:%d/", port)

	for time.Now().Before(deadline) {
		resp, err := http.Get(url)
		if err == nil {
			resp.Body.Close()
			return nil
		}
		time.Sleep(50 * time.Millisecond)
	}

	return fmt.Errorf("server did not start within %v", timeout)
}

func main() {
	startTime = time.Now()

	// Connect to IPC socket
	if err := connectIPC(); err != nil {
		fmt.Fprintf(os.Stderr, "Warning: %v\n", err)
	}

	// Find a free port for the user's server
	userPort, err := findFreePort()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to find free port: %v\n", err)
		os.Exit(1)
	}

	// Start the user's server binary
	userBinary := "./user-server"
	if _, err := os.Stat(userBinary); os.IsNotExist(err) {
		fmt.Fprintf(os.Stderr, "User server binary not found: %s\n", userBinary)
		os.Exit(1)
	}

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	cmd := exec.CommandContext(ctx, userBinary)
	cmd.Env = append(os.Environ(), fmt.Sprintf("PORT=%d", userPort))
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	if err := cmd.Start(); err != nil {
		fmt.Fprintf(os.Stderr, "Failed to start user server: %v\n", err)
		os.Exit(1)
	}

	// Wait for user's server to be ready
	if err := waitForServer(userPort, 30*time.Second); err != nil {
		fmt.Fprintf(os.Stderr, "User server failed to start: %v\n", err)
		cmd.Process.Kill()
		os.Exit(1)
	}

	// Create reverse proxy to user's server
	targetURL, _ := url.Parse(fmt.Sprintf("http://127.0.0.1:%d", userPort))
	proxy := httputil.NewSingleHostReverseProxy(targetURL)

	// Customize the proxy director to preserve headers
	originalDirector := proxy.Director
	proxy.Director = func(req *http.Request) {
		originalDirector(req)
		// Preserve the original Host header
		if host := req.Header.Get("X-Forwarded-Host"); host != "" {
			req.Host = host
		}
	}

	// The port we'll listen on (Vercel will route traffic here)
	listenPort := 3000

	// Create HTTP server with IPC-aware handler
	server := &http.Server{
		Addr: fmt.Sprintf("127.0.0.1:%d", listenPort),
		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Handle Vercel health check
			if r.URL.Path == "/_vercel/ping" {
				w.WriteHeader(http.StatusOK)
				w.Write([]byte("OK"))
				return
			}

			// Extract Vercel internal headers
			invocationID := r.Header.Get("X-Vercel-Internal-Invocation-Id")
			requestIDStr := r.Header.Get("X-Vercel-Internal-Request-Id")
			requestID, _ := strconv.ParseUint(requestIDStr, 10, 64)

			// Remove internal headers before forwarding
			for key := range r.Header {
				if strings.HasPrefix(strings.ToLower(key), "x-vercel-internal-") {
					r.Header.Del(key)
				}
			}

			// Forward request to user's server
			proxy.ServeHTTP(w, r)

			// Send end message via IPC
			if ipcConn != nil && invocationID != "" {
				endMsg := EndMessage{
					Type: "end",
					Payload: EndPayload{
						Context: RequestContext{
							InvocationID: invocationID,
							RequestID:    requestID,
						},
					},
				}
				sendIPCMessage(endMsg)
			}
		}),
	}

	// Send server-started IPC message
	initDuration := int(time.Since(startTime).Milliseconds())
	startMsg := StartMessage{
		Type: "server-started",
		Payload: StartPayload{
			InitDuration: initDuration,
			HTTPPort:     listenPort,
		},
	}

	if err := sendIPCMessage(startMsg); err != nil {
		fmt.Fprintf(os.Stderr, "Warning: Failed to send IPC start message: %v\n", err)
	} else {
		ipcReady = true
	}

	// If no IPC, print the port for local development
	if ipcConn == nil {
		fmt.Printf("Server listening on port %d (proxying to user server on port %d)\n", listenPort, userPort)
	}

	// Start the proxy server
	if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
		fmt.Fprintf(os.Stderr, "Server error: %v\n", err)
		os.Exit(1)
	}

	// Clean up
	cmd.Process.Kill()
}

// Ensure we close IPC connection on exit
func init() {
	// Note: Using a deferred cleanup in main() is preferred,
	// but this provides a fallback
}

File: packages/ocaml/vc-init.ml
(* vc-init.ml - Bootstrap wrapper for standalone OCaml servers on Vercel

   The bootstrap:
   1. Connects to VERCEL_IPC_PATH Unix socket
   2. Starts the user's server on an internal port
   3. Sends "server-started" IPC message
   4. Reverse proxies requests to user's server
   5. Handles /_vercel/ping health check
   6. Sends "end" IPC message after each request
*)

open Unix

(* IPC state *)
let ipc_socket : file_descr option ref = ref None
let start_time = ref 0.0

(* JSON helpers - minimal, no external deps *)
let json_string s =
  let buf = Buffer.create (String.length s + 2) in
  Buffer.add_char buf '"';
  String.iter (fun c ->
    match c with
    | '"' -> Buffer.add_string buf "\\\""
    | '\\' -> Buffer.add_string buf "\\\\"
    | '\n' -> Buffer.add_string buf "\\n"
    | '\r' -> Buffer.add_string buf "\\r"
    | '\t' -> Buffer.add_string buf "\\t"
    | c -> Buffer.add_char buf c
  ) s;
  Buffer.add_char buf '"';
  Buffer.contents buf

let json_int n = string_of_int n

let json_obj pairs =
  "{" ^ (String.concat ", " (List.map (fun (k, v) ->
    Printf.sprintf "\"%s\": %s" k v) pairs)) ^ "}"

(* Send IPC message (JSON + null byte) *)
let send_ipc_message json =
  match !ipc_socket with
  | None -> ()
  | Some fd ->
    let msg = json ^ "\x00" in
    let len = String.length msg in
    let rec write_all offset =
      if offset < len then
        let written = write_substring fd msg offset (len - offset) in
        write_all (offset + written)
    in
    try write_all 0 with _ -> ()

(* Connect to IPC socket *)
let connect_ipc () =
  match Sys.getenv_opt "VERCEL_IPC_PATH" with
  | None -> () (* Running locally, no IPC *)
  | Some path ->
    let fd = socket PF_UNIX SOCK_STREAM 0 in
    connect fd (ADDR_UNIX path);
    ipc_socket := Some fd

(* Find a free port *)
let find_free_port () =
  let fd = socket PF_INET SOCK_STREAM 0 in
  setsockopt fd SO_REUSEADDR true;
  bind fd (ADDR_INET (inet_addr_loopback, 0));
  let port = match getsockname fd with
    | ADDR_INET (_, p) -> p
    | _ -> failwith "unexpected address"
  in
  close fd;
  port

(* Wait for server to be ready *)
let rec wait_for_server port timeout =
  if timeout <= 0.0 then failwith "Server timeout"
  else try
    let fd = socket PF_INET SOCK_STREAM 0 in
    connect fd (ADDR_INET (inet_addr_loopback, port));
    close fd
  with _ ->
    Unix.sleepf 0.05;
    wait_for_server port (timeout -. 0.05)

(* Read a line from input channel, handling \r\n *)
let read_http_line ic =
  let line = input_line ic in
  if String.length line > 0 && line.[String.length line - 1] = '\r' then
    String.sub line 0 (String.length line - 1)
  else
    line

(* HTTP proxy implementation *)
let proxy_request client_fd user_port =
  let client_in = in_channel_of_descr client_fd in
  let client_out = out_channel_of_descr client_fd in

  (* Read request line *)
  let request_line = read_http_line client_in in

  (* Parse headers, extract Vercel internal headers *)
  let invocation_id = ref "" in
  let request_id = ref "" in
  let headers = Buffer.create 1024 in
  let content_length = ref 0 in

  let rec read_headers () =
    let line = read_http_line client_in in
    if line = "" then ()
    else begin
      let lower = String.lowercase_ascii line in
      (* Extract and remove internal headers *)
      if String.length lower > 35 &&
         String.sub lower 0 35 = "x-vercel-internal-invocation-id: " then
        invocation_id := String.trim (String.sub line 35 (String.length line - 35))
      else if String.length lower > 31 &&
              String.sub lower 0 31 = "x-vercel-internal-request-id: " then
        request_id := String.trim (String.sub line 31 (String.length line - 31))
      else if not (String.length lower > 18 &&
                   String.sub lower 0 18 = "x-vercel-internal-") then begin
        Buffer.add_string headers (line ^ "\r\n");
        if String.length lower > 16 &&
           String.sub lower 0 16 = "content-length: " then
          content_length := int_of_string (String.trim
            (String.sub line 16 (String.length line - 16)))
      end;
      read_headers ()
    end
  in
  read_headers ();

  (* Check for health check *)
  let parts = String.split_on_char ' ' request_line in
  let path = if List.length parts >= 2 then List.nth parts 1 else "/" in

  if path = "/_vercel/ping" then begin
    output_string client_out "HTTP/1.1 200 OK\r\nContent-Length: 2\r\n\r\nOK";
    flush client_out
  end else begin
    (* Proxy to user's server *)
    let user_fd = socket PF_INET SOCK_STREAM 0 in
    connect user_fd (ADDR_INET (inet_addr_loopback, user_port));
    let user_in = in_channel_of_descr user_fd in
    let user_out = out_channel_of_descr user_fd in

    (* Forward request *)
    output_string user_out (request_line ^ "\r\n");
    output_string user_out (Buffer.contents headers);
    output_string user_out "\r\n";
    if !content_length > 0 then begin
      let body = Bytes.create !content_length in
      really_input client_in body 0 !content_length;
      output_bytes user_out body
    end;
    flush user_out;

    (* Forward response *)
    let rec copy () =
      let buf = Bytes.create 4096 in
      let n = input user_in buf 0 4096 in
      if n > 0 then begin
        output client_out buf 0 n;
        copy ()
      end
    in
    (try copy () with End_of_file -> ());
    flush client_out;
    close user_fd;

    (* Send IPC end message *)
    if !invocation_id <> "" then begin
      let req_id = try int_of_string !request_id with _ -> 0 in
      send_ipc_message (json_obj [
        ("type", json_string "end");
        ("payload", json_obj [
          ("context", json_obj [
            ("invocationId", json_string !invocation_id);
            ("requestId", json_int req_id)
          ]);
          ("error", "null")
        ])
      ])
    end
  end

let () =
  start_time := Unix.gettimeofday ();

  (* Connect to IPC *)
  (try connect_ipc () with e ->
    prerr_endline ("Warning: IPC connect failed: " ^ Printexc.to_string e));

  (* Find free port for user's server *)
  let user_port = find_free_port () in

  (* Start user's server *)
  if not (Sys.file_exists "./user-server") then begin
    prerr_endline "Error: user-server not found";
    exit 1
  end;

  let env = Array.append (Unix.environment ())
    [| Printf.sprintf "PORT=%d" user_port |] in
  let pid = create_process_env "./user-server" [| "./user-server" |] env
    stdin stdout stderr in

  (* Wait for ready *)
  (try wait_for_server user_port 30.0 with e ->
    prerr_endline ("Server failed to start: " ^ Printexc.to_string e);
    kill pid Sys.sigkill;
    exit 1);

  (* Send server-started *)
  let init_duration = int_of_float
    ((Unix.gettimeofday () -. !start_time) *. 1000.0) in
  send_ipc_message (json_obj [
    ("type", json_string "server-started");
    ("payload", json_obj [
      ("initDuration", json_int init_duration);
      ("httpPort", json_int 3000)
    ])
  ]);

  if !ipc_socket = None then
    Printf.printf "Listening on :3000 (proxy to :%d)\n%!" user_port;

  (* Start server *)
  let server_fd = socket PF_INET SOCK_STREAM 0 in
  setsockopt server_fd SO_REUSEADDR true;
  bind server_fd (ADDR_INET (inet_addr_loopback, 3000));
  listen server_fd 128;

  (* Accept loop *)
  while true do
    let client_fd, _ = accept server_fd in
    (try proxy_request client_fd user_port with e ->
      prerr_endline ("Request error: " ^ Printexc.to_string e));
    close client_fd
  done
